plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixClient.java

```diff
@@ -265,6 +265,60 @@ public ResultSet executeQuery()
         };
     }
 
+    public PreparedStatement prepareStatement(
+            ConnectorSession session,
+            Connection connection,
+            JdbcTableHandle table,
+            List<JdbcColumnHandle> columns,
+            Optional<JdbcSplit> split)
+            throws SQLException
+    {
+        PreparedQuery preparedQuery = prepareQuery(
+                session,
+                connection,
+                table,
+                Optional.empty(),
+                columns,
+                ImmutableMap.of(),
+                split);
+        return new QueryBuilder(this).prepareStatement(session, connection, preparedQuery);
+    }
+
+    @Override
+    public boolean supportsTopN(ConnectorSession session, JdbcTableHandle handle, List<JdbcSortItem> sortOrder)
+    {
+        return true;
+    }
+
+    @Override
+    protected Optional<TopNFunction> topNFunction()
+    {
+        return Optional.of((query, sortItems, limit) -> {
+            // TODO: Remove when this is fixed in Phoenix
+            // Phoenix does deal well with large LIMITs with ORDER BY
+            // when there is no index matching the sort order
+            // See https://issues.apache.org/jira/browse/PHOENIX-6436
+            if (limit > MAX_LIMIT) {
+                return query;
+            }
+            String orderBy = sortItems.stream()
+                    .map(sortItem -> {
+                        String ordering = sortItem.getSortOrder().isAscending() ? "ASC" : "DESC";
+                        String nullsHandling = sortItem.getSortOrder().isNullsFirst() ? "NULLS FIRST" : "NULLS LAST";
+                        return format("%s %s %s", quoted(sortItem.getColumn().getColumnName()), ordering, nullsHandling);
+                    })
+                    .collect(joining(", "));
+
+            return format("%s ORDER BY %s OFFSET 0 ROWS FETCH NEXT %s ROWS ONLY", query, orderBy, limit);
```

Comment:
The code below this part looks exactly like `TopNFunction.sqlStandard`.
Would something like this work?

```suggestion
            if (limit > MAX_LIMIT) {
                return query;
            }
            return TopNFunction.sqlStandard(this::quoted).apply(query, sortItems, limit);
```

Reply:
Oh yes. Totally! Was musing about how to best do this.

Reply:
Done